### PR TITLE
nixos/river: init module

### PR DESCRIPTION
**Pull request by GaetanLepage**
###### Description of changes

This adds a module for the river wayland compositor.

Maintainers of the `river` package on nixpkgs: @fortuneteller2k @adamcstephens @rodrgz 

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/programs/river.nix [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        for a list of useful software.
+      '';
+    };
+
+  };
+
+  config = mkIf cfg.enable {
+    environment.systemPackages = optional (cfg.package != null) cfg.package ++ cfg.extraPackages;
+
+    security = {
```
Comment (RaitoBezarius) [inline nixos/modules/programs/river.nix:47]: This is a copy paste of sway module, I feel like this should be factored out

Comment (GaetanLepage) [reply nixos/modules/programs/river.nix:47]: Indeed it is. What would be a good way to factor it out ?
Add some kind of `wayland-window-manager.nix` file that would get imported by both `sway` and `river` ?

Comment (GaetanLepage) [reply nixos/modules/programs/river.nix]: Done ! Does it seem fine for you like this ?

File: nixos/modules/programs/wayland/wayland-session.nix [hunk 2]
```diff
@@ -0,0 +1,26 @@
+{ lib, ... }: with lib; {
+
```
Comment (RaitoBezarius) [inline nixos/modules/programs/wayland/wayland-session.nix:2]:
```suggestion
```

File: nixos/modules/programs/wayland/wayland-session.nix [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    hardware.opengl.enable = mkDefault true;
+
+    fonts.enableDefaultFonts = mkDefault true;
+
+    programs = {
+      dconf.enable = mkDefault true;
+      xwayland.enable = mkDefault true;
+    };
+
```
Comment (RaitoBezarius) [inline nixos/modules/programs/wayland/wayland-session.nix:16]:
```suggestion
```

File: nixos/modules/programs/wayland/wayland-session.nix [hunk 4]
```diff
@@ -0,0 +1,26 @@
+{ lib, ... }: with lib; {
+
+    security = {
+      polkit.enable = true;
+      pam.services.swaylock = {};
+    };
+
+    hardware.opengl.enable = mkDefault true;
+
```
Comment (RaitoBezarius) [inline nixos/modules/programs/wayland/wayland-session.nix:9]:
```suggestion
```

File: nixos/modules/programs/wayland/wayland-session.nix [hunk 5]
```diff
@@ -0,0 +1,26 @@
+{ lib, ... }: with lib; {
```
Comment (RaitoBezarius) [inline nixos/modules/programs/wayland/wayland-session.nix:1]:
```suggestion
{ lib, pkgs, ... }: with lib; {
```

**Review by RaitoBezarius — CHANGES_REQUESTED** [review]
This was not tested, it does not work.

**Comment by GaetanLepage** [comment]
Thanks for your review @RaitoBezarius !
I submitted the requested changes. Please, tell me if I should correct more things.

File: nixos/doc/manual/release-notes/rl-2305.section.md [hunk 6]
```diff
@@ -50,6 +50,8 @@ In addition to numerous new and upgraded packages, this release has the followin
 
 - [Cloudlog](https://www.magicbug.co.uk/cloudlog/), a web-based Amateur Radio logging application. Available as [services.cloudlog](#opt-services.cloudlog.enable).
 
+- [river](https://github.com/riverwm/river), A dynamic tiling wayland compositor. Available as [programs.river](#opt-programs.river.enable).
```
Comment (RaitoBezarius) [inline nixos/doc/manual/release-notes/rl-2305.section.md:53]: You have to move this in 2311 now

Comment (RaitoBezarius) [reply nixos/doc/manual/release-notes/rl-2305.section.md:53]: We branched off 23.05 2 days ago

Comment (GaetanLepage) [reply nixos/doc/manual/release-notes/rl-2305.section.md]: Done !

**Review by RaitoBezarius — APPROVED** [review]
Tested for the sway part.

**Pull request merged**